### PR TITLE
provider: allow fallback to installed packages

### PR DESCRIPTION
**Pull request by abn**
Split out from https://github.com/python-poetry/poetry/pull/5450

### CODE REVIEW

File: src/poetry/puzzle/provider.py [hunk 1]
```diff
@@ -478,15 +516,25 @@ def complete_package(self, package: DependencyPackage) -> DependencyPackage:
...
+                        repository=package.dependency.source_name,
+                    ),
+                )
+            except PackageNotFound as e:
+                for pkg in self.search_for_installed_packages(package.dependency):
+                    package = DependencyPackage(
+                        package.dependency,
+                        pkg,
+                    )
+                    break
```
Comment (neersighted) [inline src/poetry/puzzle/provider.py:535]: I'm not quite sure what this break is for -- if we just intend to get the first package in the list, a call to `next()` might be more explicit.

Comment (abn) [reply src/poetry/puzzle/provider.py:535]: More to trigger the else block in the for loop, we can also do `next` and then catch the `StopIteration`. For me this is a bit cleaner; but I am happy to switch to `next` too.

Comment (neersighted) [reply src/poetry/puzzle/provider.py:535]: Ah, I missed the for-else -- I think I would prefer a next/StopIteration as the pattern is less likely to get lost in indentation noise (like just happened to me) and the intent is more explicit.

Comment (abn) [reply src/poetry/puzzle/provider.py:535]: Fine! Here I was ... all excited I could use a `for/else` :D

**Review by neersighted — APPROVED** [review]

**Pull request merged**